src/units/rtr.rs

```diff
@@ -466,47 +584,195 @@ impl PayloadUpdate for TargetUpdate {
+        MetricType::Text, MetricUnit::Info
+    );
+    const BYTES_READ_METRIC: Metric = Metric::new(
+        "bytes_read", "the number of bytes read",
+        MetricType::Counter, MetricUnit::Total,
+    );
+    const BYTES_WRITTEN_METRIC: Metric = Metric::new(
+        "bytes_written", "the number of bytes written",
+        MetricType::Counter, MetricUnit::Total,
+    );
+
+    const ISO_DATE: &'static [chrono::format::Item<'static>] = &[
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Year, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal("-"),
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Month, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal("-"),
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Day, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal("T"),
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Hour, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal(":"),
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Minute, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal(":"),
+        chrono::format::Item::Numeric(
+            chrono::format::Numeric::Second, chrono::format::Pad::Zero
+        ),
+        chrono::format::Item::Literal("Z"),
+    ];
 }
 
 impl metrics::Source for RtrMetrics {
     fn append(&self, unit_name: &str, target: &mut metrics::Target)  {
         self.gate.append(unit_name, target);
+
+        let session = self.session.load(atomic::Ordering::Relaxed);
+        if session != u32::MAX {
+            target.append_simple(
+                &Self::SESSION_METRIC, Some(unit_name), session
+            );
+        }
+
+        let serial = self.serial.load(atomic::Ordering::Relaxed);
+        if serial != u32::MAX {
+            target.append_simple(
+                &Self::SERIAL_METRIC, Some(unit_name), serial
+            )
+        }
+
+        let updated = self.updated.load(atomic::Ordering::Relaxed);
+        if updated != i64::MIN {
+            let updated = Utc.timestamp(updated, 0);
+            let ago = Utc::now().signed_duration_since(updated);
+            target.append_simple(
+                &Self::UPDATED_AGO_METRIC, Some(unit_name), ago.num_seconds()
+            );
+            target.append_simple(
+                &Self::UPDATED_METRIC, Some(unit_name),
+                updated.format_with_items(Self::ISO_DATE.iter())
+            );
+        }
+
+        target.append_simple(
+            &Self::BYTES_READ_METRIC, Some(unit_name),
+            self.bytes_read.load(atomic::Ordering::Relaxed)
+        );
+        target.append_simple(
+            &Self::BYTES_WRITTEN_METRIC, Some(unit_name),
+            self.bytes_written.load(atomic::Ordering::Relaxed)
+        );
     }
 }
 
```

Comment:
I think that the subsequent atomic `.load()` would allow the `serial`, `bytes_read` and `bytes_written` to be mixed up (like, serial gets associated with some other serial's counters) if you don't set a memory fence around all the load()'s. Well, unless you're sure that only one thread writes to one `session` of course

Reply:
I think that’s the case (there’s only one future reading/writing). But what would be the correct ordering here?

Reply:
I think `Acquire` and `Release`. But that would only work if you do a `compare_exchange` (or another operation that verifies the serial number) at the end of the loads() to verify that the serial wasn't changed in the meantime (by another thread) - and to make the ordering actually do something. Or set a memory fence, using [std](https://doc.rust-lang.org/std/index.html)::[sync](https://doc.rust-lang.org/std/sync/index.html)::[atomic](https://doc.rust-lang.org/std/sync/atomic/index.html)::[fence](https://doc.rust-lang.org/std/sync/atomic/fn.fence.html#)

Reply:
Sorry, I should have looked more carefully. At that point the atomic is read, for which I think relaxed is fine – if the metric is a bit behind, that doesn’t really quite matter.

Reply:
Ok, All's fine we found out. `bytes_read|written` are incrementing totals, so they're not tied to the serial number. So the worst case is that the serial number might be ahead of the actual reported bytes read|written.